### DISCUSSION TITLE
Rationale for no leading space in `//nolint:mylinter`?

### DISCUSSION THREAD

**Discussion by sebastien-rosset** [golangci/golangci-lint#3271] [Q&A] [answered]
The `nolintlint` linter requires there is no leading space before `nolint`. For example:
```diff
- // nolint: gocyclo
+ //nolint: gocyclo
func myBigFunc() error {
```

The rule was introduced in https://github.com/golangci/golangci-lint/pull/837. Then it became a hard rule, as discussed briefly here: https://github.com/golangci/golangci-lint/issues/1658#issuecomment-1184372477

However, this seems to go against the new go formatting rules introduced in `gofmt` version 1.19. For example, if the `//nolint:` comment is above a function, then `gofmt` will automatically add a leading space. The following code will make `nolintlint` happy:

```go
//nolint: gocyclo
func myBigFunc() error {
```

But if you run `gofmt` version 1.19, the code will be reformatted as shown below, and now `nolintlint` will complain there is a leading space.
```go
// nolint: gocyclo
func myBigFunc() error {
```

**Comment by ldez** [comment, marked as answer]
Hello,

`nolint` is a directive, not a comment, the new Go fmt adds a space before a comment, not before a directive.

The option `allow-leading-space` has been dropped.

The syntax of directives is: `//[a-zA-Z]+:[a-zA-Z].*`.
The spaces around `:` and after the `//` must be removed.

I already explained that in all the linked issues of https://github.com/golangci/golangci-lint/issues/1658

**Reply by bombsimon** [reply]
To elaborate on this, the directive should be `//nolint:gocyclo` without any space and `gofmt` should then leave the "comment" as is since it will understand it's not a regular comment.

**Reply by sebastien-rosset** [reply]
That's not what I am observing with go 1.19.1 though.

The directive below "works". `gofmt` does not change the formatting.
```go
func myBigFunc() error { //nolint:gocyclo // reason why
```

If the directive is above the applicable line of code, then `gofmt` version 1.19.1 adds a leading space, and then `nolintlint` complains there is a leading space. E.g.:
```diff
- //nolint:gocyclo // reason why
+ // nolint:gocyclo // reason why
func myBigFunc() error {
```

This seems to be an open issue: https://github.com/golang/go/issues/43776
If you have a code base with a lot of such directives, then when you upgrade to go 1.19, lots of leading spaces are added by `gofmt` and at the same time `nolintlint` complains about the leading space.

**Reply by bombsimon** [reply]
I cannot reproduce this, this is the result for me.

```sh
$ go1.19.1 version
go version go1.19.1 linux/amd64
```

```sh
$ cat main.go
package main

//nolint:gocyclo // reason why
func main() {}

//nolint: gocyclo // reason why
func notmain() {}
```

```sh
$ git diff
diff --git a/main.go b/main.go
index e3afcb2..d4e6c49 100644
--- a/main.go
+++ b/main.go
@@ -3,5 +3,5 @@ package main
 //nolint:gocyclo // reason why
 func main() {}

-//nolint: gocyclo // reason why
+// nolint: gocyclo // reason why
 func notmain() {}
```

**Reply by ldez** [reply]
I have the behavior as @bombsimon.

The directive formal definition: https://github.com/golang/go/issues/37974#issuecomment-611088006

**Reply by sebastien-rosset** [reply]
Presumably you typed `git diff` after typing `gofmt -w .`, right? (with gofmt 1.19.1)

**Reply by bombsimon** [reply]
Yes, I typed them in the order I pasted them. The diff was to show what `gofmt` changed which was only the directive with a space after the colon.

**Reply by ldez** [reply]
yes
